assets/revisions/rev_lcq797n5ryxk_migrate_sample_files.py

```diff
@@ -0,0 +1,334 @@
+        return True
+
+    names = [file["name"] for file in files]
+
+    return all(name in {"reads_1.fq.gz", "reads_2.fq.gz"} for name in names)
+
+
+def join_legacy_read_paths(data_path: Path, paired: bool, sample_id: int):
+    """Create a list of paths for the read files associated with the `sample`.
+
+    :param data_path: the location of the data directory
+    :param paired: whether paired or single end reads are used
+    :param sample_id: the unique identifier of the sample
+    :return: a list of sample read paths
+    """
+    sample_path = data_path / "samples" / sample_id
+
+    if paired:
+        return [
+            sample_path / "reads_1.fastq",
+            sample_path / "reads_2.fastq",
+        ]
+
+    return [sample_path / "reads_1.fastq"]
+
+
+async def compress_sample_reads(
+    db: AsyncIOMotorDatabase,
+    data_path: Path,
+    sample_files: dict[str, any],
+    paired: bool,
+    sample_id: int,
+):
+    """Compress the reads for one legacy samples.
+
+    :param db: the application database object
+    :param data_path: the location of the data directory
+    :param files: the read files to be compressed
+    :param paired: whether paired or single end reads are used
+    :param sample_id: the unique identifier of the sample
+
+    """
+    paths = join_legacy_read_paths(data_path, paired, sample_id)
+
+    files = []
+
+    for i, path in enumerate(paths):
+        target_filename = (
+            "reads_1.fq.gz" if "reads_1.fastq" in str(path) else "reads_2.fq.gz"
+        )
+
+        target_path = data_path / "samples" / sample_id / target_filename
+
+        await to_thread(compress_file, path, target_path, 1)
+
+        stats = await to_thread(file_stats, target_path)
+
+        files.append(
+            {
+                "name": target_filename,
+                "download_url": f"/download/samples/{sample_id}/{target_filename}",
+                "size": stats["size"],
+                "raw": False,
+                "from": sample_files[i]["from"],
+            },
+        )
+
+    await db.samples.update_one(
+        {"_id": sample_id},
+        {"$set": {"files": files, "is_compressed": True}},
+    )
+
+    for path in paths:
+        await to_thread(os.remove, path)
+
+    return files
+
+
+async def upgrade(ctx: MigrationContext):
+    async for sample in ctx.mongo.samples.find({"files": {"$exists": True}}):
+        if "is_legacy" not in sample:
+            sample["is_legacy"] = check_is_legacy(sample)
+            ctx.mongo.samples.update_one(
+                {"_id": sample["_id"]},
+                {"$set": {"is_legacy": sample["is_legacy"]}},
+            )
+
+        if not check_is_compressed(sample):
+            sample["files"] = await compress_sample_reads(
+                ctx.mongo,
+                ctx.data_path,
+                sample["files"],
+                sample["paired"],
+                sample["_id"],
+            )
```

Comment:
I think the Mongo changes here should also happen in the below Mongo transaction.

Reply:
I think this probably makes sense, but I need to be careful with getting this right. The key advantage to the way it is done now is that the files will never be deleted before the new `file` location is saved to a database. With merging them I potentially lose track of which samples have been compressed during this migration and which were already compressed. That said, I think blindly attempting to delete the uncompressed reads for each sample shouldn't cause any problems